File: 1512-DesignUndergroundSystem/1512-DesignUndergroundSystem.java
// Last updated: 3/26/2026, 1:55:04 AM
class UndergroundSystem {
    Map<Integer,Pair<String,Integer>> checkIn;
    Map<String, Pair<Integer,Integer>> avg;
    
    public UndergroundSystem() {
        checkIn = new HashMap<>();
        avg = new HashMap<>();
    }
    
    public void checkIn(int id, String stationName, int t) {
        checkIn.put(id,new Pair(stationName,t));
    }
    
    public void checkOut(int id, String stationName, int t) {
        Pair<String,Integer> source = checkIn.get(id);
        
        String stationKey = source.getKey() + "#" + stationName;
        int totalTime = t-source.getValue();
        if(!avg.containsKey(stationKey)){
            avg.put(stationKey,new Pair(totalTime,1));
        }
        else{
            Pair<Integer,Integer> old_value = avg.get(stationKey);
            Pair<Integer,Integer> new_value = new Pair(old_value.getKey() + totalTime,old_value.getValue()+1);
            avg.put(stationKey,new_value);
        }
    }
    
    public double getAverageTime(String startStation, String endStation) {
        String stationKey = startStation+"#"+endStation;
        Pair<Integer,Integer> timeValue = avg.get(stationKey);
        return (double)timeValue.getKey()/timeValue.getValue();
    }
}

/**
 * Your UndergroundSystem object will be instantiated and called as such:
 * UndergroundSystem obj = new UndergroundSystem();
 * obj.checkIn(id,stationName,t);
 * obj.checkOut(id,stationName,t);
 * double param_3 = obj.getAverageTime(startStation,endStation);
 
 checkinMap - [ID,Pair<sourceStation,time)]
 checkout - 
 Source station, start time, end time = end time- start time = totalTime,
 
 AverageTime - [SourceStation-DestinationStation, <TotalTime,count>]
 
 
 */

File: 1512-DesignUndergroundSystem/1512-DesignUndergroundSystem.cs
// Last updated: 3/26/2026, 1:55:02 AM
public class UndergroundSystem {
    Dictionary<int, (string,int)> checkin;
    Dictionary<string,(int count,double avg)> averageTimes;

    public UndergroundSystem() {
        checkin = new();
        averageTimes = new();
    }
    
    public void CheckIn(int id, string stationName, int t) {
        if(!checkin.ContainsKey(id))
        {
            checkin.Add(id,(stationName,t));
        }
        checkin[id] = (stationName,t);
    }
    
    public void CheckOut(int id, string stationName, int t) {
        var startStation = checkin[id];

        string route = checkin[id].Item1+"->"+stationName;
        int time = t - startStation.Item2;

        if(!averageTimes.ContainsKey(route))
        {
            averageTimes.Add(route,(0,0));
        }
        var calc = averageTimes[route];
        averageTimes[route] = (calc.count+1, (double)(((double)calc.count*calc.avg)+time)/(calc.count+1));
    }
    
    public double GetAverageTime(string startStation, string endStation) {
        return averageTimes[startStation+"->"+endStation].avg;
    }
}

/**
 * Your UndergroundSystem object will be instantiated and called as such:
 * UndergroundSystem obj = new UndergroundSystem();
 * obj.CheckIn(id,stationName,t);
 * obj.CheckOut(id,stationName,t);
 * double param_3 = obj.GetAverageTime(startStation,endStation);
 */